api/units.go

```diff
@@ -101,15 +101,33 @@ func (ur *unitsResource) set(rw http.ResponseWriter, req *http.Request, item str
 		return
 	}
 
+	newUnit := false
 	if eu == nil {
 		if len(su.Options) == 0 {
 			err := errors.New("unit does not exist and options field empty")
 			sendError(rw, http.StatusConflict, err)
+			return
 		} else if err := ValidateOptions(su.Options); err != nil {
 			sendError(rw, http.StatusBadRequest, err)
+			return
 		} else {
-			ur.create(rw, su.Name, &su)
+			newUnit = true
 		}
+	} else if eu.Name == su.Name && len(su.Options) > 0 {
+		// There is already a unit with the same name
+		// check the hashes if they do not match then we will
+		// create a new unit with the same name and later
+		// the job will be updated to this new unit.
```

Comment:
> Well, the commit messages and code changes seem to suggest that just submitting a unit with the same name will replace the existing one; which was my original understanding of how this is supposed to work.

It's clearly commented:
https://github.com/coreos/fleet/pull/1509/commits/0e675d9dca2ebd42d296bde75466e2e49b6daa47#diff-273bb648504b6b04dd506aa86e10e573R118

>  It's the code comment here that threw me off. I think the wording is extremely confusing -- especially the ominous "later"... I suspect what you are actually trying to say is that the reconciler will update the running jobs once the unit file has been replaced in the registry -- but that's not clear at all.

Oh sorry @antrik it seems that you are still with the old idea, yep! so yeh it is still not clear! 

So by later we just mean update the job key to point to the new unit that we have to create anyway, and which will be in the inactive state, which in turn will trigger a unit stop or unload for the old unit if necessary, and make the new unit in the inactive state. So we don't wait for the reconciler at all which will do the same job but might take longer and needs code..., we don't trigger that path, and we don't add more code to update desired and current hashes inside the job keys. All we do: is set job keys manually to the new unit (again that we have to create manually anyway...) and put it into the "inactive" state, and everything else continue to work automatically... and we use the same wait for block logic for inactive, load or start states... and current fleet transition design to the final state is what helped here.

Thank you!